# Introduction to Optuna + RAPIDS

Hyperparameter optimization is the task of picking values for the hyperparamters of the model to yield optimal results. This can help boost the model accuracy greatly, but can be resource-intensive and is often not widely used for practical purposes. Let's take a look at how we can use Optuna and RAPIDS to make HPO less time-consuming.


## Optuna
[Optuna](https://optuna.readthedocs.io/en/stable/) is a lightweight framework for automatic hyperparameter optimization. It provides a define-by-run API, which makes it easy to adapt to any already existing code that we have and enables high modularity and the flexibility to construct hyperparameter spaces dynamically. By simply wrapping the objective function with Optuna can help perform a parallel-distributed HPO search over a search space. As we'll see in this notebook.


## RAPIDS
RAPIDS framework provides a library suite that can execute end-to-end data science pipelines entirely on GPUs. One of the libraries in this framework is cuML, which contains various Machine Learning algorithms that take advantage of GPU to run. You can learn more about RAPIDS [here](https://rapids.ai/about.html).


In this notebook, we'll use Airlines dataset (20M rows) to predict if a flight will be delayed or not. We'll explore how to use Optuna with RAPIDS and the speedups that we can achieve with the integration of these.

In [1]:
# Run this cell to install optuna and mlflow
# !pip install optuna
# !pip install mlflow

In [2]:
import random
import time
from contextlib import contextmanager

import cudf
import cuml
import dask_cudf
import mlflow
import numpy as np
import optuna
import pandas as pd
import sklearn
from cuml.dask.common import utils as dask_utils
from cuml.metrics import accuracy_score
from cuml.preprocessing.model_selection import train_test_split
from dask.distributed import Client, wait, performance_report
from joblib import parallel_backend, Parallel, delayed

from sklearn.datasets import load_iris

from dask_cuda import LocalCUDACluster

/opt/conda/envs/rapids/lib/python3.6/site-packages/treelite/gallery/__init__.py:7: FutureWarning: treelite.gallery.sklearn has been moved to treelite.sklearn. treelite.gallery.sklearn will be removed in version 1.1.
  FutureWarning)
/opt/conda/envs/rapids/lib/python3.6/site-packages/treelite/gallery/sklearn/__init__.py:9: FutureWarning: treelite.gallery.sklearn has been moved to treelite.sklearn. treelite.gallery.sklearn will be removed in version 1.1.
  FutureWarning)


In [3]:
# Helper function for timing blocks of code.
@contextmanager
def timed(name):
    t0 = time.time()
    yield
    t1 = time.time()
    print("..%-24s:  %8.4f" % (name, t1 - t0))

## Set up CUDA Cluster

We start a local cluster and keep it ready for running distributed tasks with dask.

[LocalCUDACluster](https://github.com/rapidsai/dask-cuda) launches one Dask worker for each GPU in the current systems. It's developed as a part of the RAPIDS project. Learn More:
- [Setting up Dask](https://docs.dask.org/en/latest/setup.html)
- [Dask Client](https://distributed.dask.org/en/latest/client.html)

In [4]:
# This will use all GPUs on the local host by default
cluster = LocalCUDACluster(threads_per_worker=1, ip="", dashboard_address="8002")
c = Client(cluster)

# Query the client for all connected workers
workers = c.has_what().keys()
n_workers = len(workers)
n_streams = 8 # Performance optimization
c

<Client: 'tcp://172.17.0.2:42787' processes=2 threads=2, memory=49.16 GB>

## Loading the data

### Data Acquisition

We'll use the cell below to download the data. The `file_name` specifies which of the two available files - airline_small.parquet (smaller file) and airline_20000000.parquet we want to use. By default, we use the smaller file, but the benchmarks were run with the larger file. You are free to change it for experimentation.

Run the cell below just once in the local machine to acquire data. Comment out for future runs.

In [5]:
# # Read above instructions - RUN ONLY ONCE

# from urllib.request import urlretrieve
# import os

# file_name = 'airline_small.parquet' # NOTE: Change to airline_20000000.parquet to use a larger dataset

# data_dir = "/local/path/to/file/" # NOTE: Change to a local path where you want to save the file
# INPUT_FILE = os.path.join(data_dir, file_name)

# if os.path.isfile(INPUT_FILE):
#         print(f" > File already exists. Ready to load at {INPUT_FILE}")
# else:
#     # Ensure folder exists
#     os.makedirs(data_dir, exist_ok=True)
        
# url = "https://rapidsai-cloud-ml-sample-data.s3-us-west-2.amazonaws.com/" + file_name

# urlretrieve(url= url,filename=INPUT_FILE)

# print("Completed!")

Select the `N_TRIALS` for the number of runs of HPO trials. 

We will now, load the data from `INPUT_FILE`

In [6]:
N_TRIALS = 20

INPUT_FILE = "/home/hyperopt/hyperopt/data/air_par.parquet"
df = cudf.read_parquet(INPUT_FILE)
X, y = df.drop(["ArrDelayBinary"], axis=1), df["ArrDelayBinary"].astype('int32')

# Training and Evaluation

Here, we define `train_and_eval` function which simply fits a RandomForestClassifier (with`max_depth` and `n_estimators`) on the passed `X_param`, `y_param`. This function should look very similar for any ML workflow. We'll use this function within the Optuna `objective` function to show how easily we can fit an existing workflow into the Optuna work. 

In [7]:
def train_and_eval(X_param, y_param, max_depth=16, n_estimators=100):
    """
        Splits the given data into train and test split to train and evaluate the model
        for the params parameters.
        
        Params
        ______
        
        X_param:  DataFrame. 
                  The data to use for training and testing. 
        y_param:  Series. 
                  The label for training
        max_depth, n_estimators: The values to use for max_depth and n_estimators for RFC.
                                 Defaults to 16 and 100 (the defaults for the classifiers used)
                   
        Returns
        score: Accuracy score of the fitted model
    """

    X_train, X_valid, y_train, y_valid = train_test_split(X_param, y_param, random_state=77)
    classifier = cuml.ensemble.RandomForestClassifier(max_depth=max_depth,
                     n_estimators=n_estimators)
    classifier.fit(X_train, y_train)
    y_pred = classifier.predict(X_valid)
    score = accuracy_score(y_valid, y_pred)
    return score

For a baseline number, let's see what the default performance of RFC is. Note the defauly values for `max_depth` = 16 and `n_estimators` = 100; we pass these to the `train_and_eval` function.

In [8]:
print("Score with default parameters : ",train_and_eval(X, y, max_depth=16, n_estimators=100))

Score with default parameters :  0.837492823600769


## Objective Function

The objective function will be the one we optimize in [Optuna Study](https://optuna.readthedocs.io/en/stable/reference/study.html). Objective funciton tries out different values for the parameters that we are tuning and saving the results in `study.trials_dataframes()`. 

Let's define the objective function for this HPO task by making use of the `train_and_eval()`. You can see that we simply choose a value for the parameters and call the `train_and_eval` method, making Optuna very easy to use in an existing workflow.

The objective remains constant over different [samplers](https://optuna.readthedocs.io/en/stable/reference/samplers.html), which are built-in options in Optuna to enable the selection of different sampling algorithms that optuna provides. Some of the available ones include - GridSampler, RandomSampler, TPESampler, etc. We'll use TPESampler for this demo, but feel free to try different samplers to notice the chnages in performance.

In [9]:
def objective(trial, X_param, y_param):
    max_depth = trial.suggest_int("max_depth", 10, 15)
    n_estimators = trial.suggest_int("n_estimators", 200, 700)
    score = train_and_eval(X_param, y_param, max_depth=max_depth,
                           n_estimators=n_estimators)
    return score

## HPO Trials and Study

Optuna uses [study](https://optuna.readthedocs.io/en/stable/reference/study.html) and [trials](https://optuna.readthedocs.io/en/stable/reference/trial.html) to keep track of the HPO experiments. Put simply, a trial is a single call of the objective function while a set of trials make up a study. We will pick the optimal performing trial from a study to get the best parameters that were used in that run.

We'll make use of a helper function `run_study` to help us run one multi-GPU study with a dask backend using [Joblib](https://joblib.readthedocs.io/en/latest/). We make use of `parallel_backend` from Joblib to allow the optuna studies to run in parallel on the GPUs as we make use of the dask backend. As you can observe, it accepts the client that we set up earlier and the data X and y are scattered across them - this reduces the time it takes for data transfer among the GPUs.

Optuna also requires the used of a storage to run distributed optimization runs. Learn more about what storages can be used [here](https://optuna.readthedocs.io/en/stable/tutorial/distributed.html)

In [10]:
def run_study(sampler=optuna.samplers.TPESampler(),
              study_name="Optuna-MultiGPU",
              callbacks=None):
    
    with timed(study_name):
        study = optuna.create_study(sampler=sampler,
                                    study_name=study_name,
                                    storage="sqlite:///_"+study_name+".db",
                                    direction="maximize",
                                    load_if_exists=True)
        
        with parallel_backend("dask", n_jobs=n_workers, client=c, scatter=[X,y]):
            study.optimize(lambda trial: objective(trial, X, y),
                           n_trials=N_TRIALS,
                           n_jobs=n_workers,
                           callbacks=callbacks)
    print("Number of finished trials: ", len(study.trials))
    print("Best trial:")
    trial = study.best_trial
    print("  Value: ", trial.value)
    print("  Params: ")
    for key, value in trial.params.items():
        print("    {}: {}".format(key, value))
    return study

In [11]:
name = "optuna-joblib-dask-backend"
with performance_report(filename=name+"-dask_report.html"):
    study_tpe = run_study(optuna.samplers.TPESampler(),
                          study_name=name)

[I 2020-07-07 00:00:34,923] A new study created with name: optuna-joblib-dask-backend


..optuna-joblib-dask-backend:  526.3728
Number of finished trials:  20
Best trial:
  Value:  0.8430033922195435
  Params: 
    max_depth: 15
    n_estimators: 502


#### Now, how long would it take without dask or multiprocessing with joblib?

In [13]:
name = "optuna-simple"
with timed("no-dask-no-joblib"):
    study = optuna.create_study(sampler=optuna.samplers.TPESampler(),
                                study_name=name,
                                storage="sqlite:///_"+name+".db",
                                direction="maximize",
                                load_if_exists=True)
    study.optimize(lambda trial: objective(trial, X, y), n_trials=N_TRIALS)

[I 2020-07-07 00:24:35,853] A new study created with name: optuna-simple
[I 2020-07-07 00:25:12,172] Finished trial#0 with value: 0.8307780027389526 with parameters: {'max_depth': 13, 'n_estimators': 514}. Best is trial#0 with value: 0.8307780027389526.
[I 2020-07-07 00:25:36,235] Finished trial#1 with value: 0.8307763934135437 with parameters: {'max_depth': 13, 'n_estimators': 343}. Best is trial#0 with value: 0.8307780027389526.
[I 2020-07-07 00:25:53,840] Finished trial#2 with value: 0.8308228254318237 with parameters: {'max_depth': 14, 'n_estimators': 221}. Best is trial#2 with value: 0.8308228254318237.
[I 2020-07-07 00:26:23,624] Finished trial#3 with value: 0.8342193961143494 with parameters: {'max_depth': 10, 'n_estimators': 640}. Best is trial#3 with value: 0.8342193961143494.
[I 2020-07-07 00:26:41,571] Finished trial#4 with value: 0.8341361880302429 with parameters: {'max_depth': 10, 'n_estimators': 379}. Best is trial#3 with value: 0.8342193961143494.
[I 2020-07-07 00:27:01

..no-dask-no-joblib       :  750.3215


Expectely it takes a lot longer to run a sequential Optuna call


# Sequential calls without Optuna

For a comparison let's try sequential calls without Optuna and it's parallel-processing support. We'll pick the same parameters as Optuna for a fair comparison - these parameters were selected by the sampling algorithm used by Optuna and is available in the `study.trials_dataframe()` for us to pick out.

In [14]:
df = study_tpe.trials_dataframe()
params_max_depth, params_n_estimators = df['params_max_depth'], df['params_n_estimators']

### Sequential call function 

For a cleaner look, let's use a function to perform sequential calls. The function basically sets the parameters to what was passed and trains and evaluates the model and returns the details of the run which can later be used to find the best performing model.

In [15]:
def seq_call(X, y, max_depth, n_estimators):
    
    score = train_and_eval(X, y, max_depth=max_depth, n_estimators = n_estimators)
    
    return score

In [16]:
name = "sequential-calls"
max_acc = -1
with timed(name):
    for i in range(N_TRIALS):
        acc = seq_call(X, y, max_depth=params_max_depth[i],
                     n_estimators=params_n_estimators[i])
        if acc > max_acc:
            max_acc = acc
            best_max_depth = params_max_depth[i]
            best_n_est = params_n_estimators[i]

print(max_acc, best_max_depth, best_n_est)

..sequential-calls        :  857.3299
0.843004584312439 15 502


Running it without any framework expectedly takes the longest to run. 

## Summarizing the timing results

| Study name | Runtime |   
|---|---|
| Optuna-Multi-GPU-TPE | 526.3728 |
| No-dask-No-Joblib | 750.3215 |
| No-Optuna-No-dask-Seq-Call | 857.3299 |

We see the Optuna with parallel runs of Single-GPU estimators yields much better results than the alternatives. This will help run more iterations of hyperparameter search and yield quicker results. This can serve as a good starting point to try out Optuna + RAPIDS to see faster HPO runs.